File: src/main/java/ch/uzh/ifi/access/repository/TaskRepository.java
package ch.uzh.ifi.access.repository;

import ch.uzh.ifi.access.model.Task;
import ch.uzh.ifi.access.projections.TaskWorkspace;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.security.access.prepost.PostAuthorize;

import java.util.Optional;

public interface TaskRepository extends JpaRepository<Task, Long> {

    @PostAuthorize("hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug) and returnObject.get().published)")
    Optional<TaskWorkspace> findByAssignment_Course_SlugAndAssignment_SlugAndSlug(String courseSlug, String assignmentSlug, String taskSlug);

    Optional<Task> getByAssignment_Course_SlugAndAssignment_SlugAndSlug(String courseSlug, String assignmentSlug, String taskSlug);

    boolean existsByAssignment_Course_SlugAndAssignment_SlugAndSlug(String courseSlug, String assignmentSlug, String taskSlug);

}

File: src/main/kotlin/ch/uzh/ifi/access/repository/TaskRepository.kt
package ch.uzh.ifi.access.repository

import ch.uzh.ifi.access.model.Task
import ch.uzh.ifi.access.projections.TaskWorkspace
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.security.access.prepost.PostAuthorize

interface TaskRepository : JpaRepository<Task?, Long?> {
    // TODO: visibility based on date
    //@PostAuthorize("hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug) and returnObject.get().published)")
    @PostAuthorize("hasRole(#courseSlug + '-assistant') or (hasRole(#courseSlug))")
    fun findByAssignment_Course_SlugAndAssignment_SlugAndSlug(
        courseSlug: String?,
        assignmentSlug: String?,
        taskSlug: String?
    ): TaskWorkspace?

    fun getByAssignment_Course_SlugAndAssignment_SlugAndSlug(
        courseSlug: String?,
        assignmentSlug: String?,
        taskSlug: String?
    ): Task?

    // Id-only resolution used by the cached TaskLookupService (hot submission path).
    @Query("select t.id from Task t where t.assignment.course.slug = ?1 and t.assignment.slug = ?2 and t.slug = ?3")
    fun findTaskIdBySlugs(courseSlug: String, assignmentSlug: String, taskSlug: String): Long?

    fun existsByAssignment_Course_SlugAndAssignment_SlugAndSlug(
        courseSlug: String?,
        assignmentSlug: String?,
        taskSlug: String?
    ): Boolean
}
